--- a/Makefile.in	2019-04-29 20:45:25.457468597 +0200
+++ b/Makefile.in	2019-05-08 18:26:50.141694821 +0200
@@ -533,9 +533,8 @@
                           python-pkg/README.rst
 
 py_pkg_lircdir = $(pkgdatadir)/python-pkg/lirc
-py_PYTHON = python-pkg/lirc/__init__.py python-pkg/lirc/database.py \
+py_PYTHON = python-pkg/lirc/__init__.py python-pkg/lirc/config.py python-pkg/lirc/database.py \
 	python-pkg/lirc/paths.py $(am__append_1)
-nodist_py_pkg_PYTHON = python-pkg/lirc/config.py
 @HAVE_PYTHON35_TRUE@py_pkg_testdir = $(pkgdatadir)/python-pkg/tests
 @HAVE_PYTHON35_TRUE@dist_py_pkg_test_DATA = python-pkg/tests/test_client.py \
 @HAVE_PYTHON35_TRUE@                          python-pkg/tests/dummy-server \
